File: app/frontend/src/app/game/components/2_navigasi_menu/2_navigasi_bawah/5_geopolitik/2_organisasi_internasional/2_organisasi_regional/logic/python/membership_movement/regional_movement_engine.py
import random
import os
import sys

# Add subfolders to path
sys.path.append(os.path.join(os.path.dirname(__file__), '1_join_logic'))
sys.path.append(os.path.join(os.path.dirname(__file__), '2_leave_logic'))

from regional_join_handler import RegionalJoinHandler
from regional_leave_handler import RegionalLeaveHandler

class RegionalMovementEngine:
    """
    Handles the movement logic for regional organizations.
    Coordination between joining and leaving handlers.
    """
    
    def __init__(self):
        self.join_handler = RegionalJoinHandler()
        self.leave_handler = RegionalLeaveHandler()

    def calculate_movement(self, country_name, current_status, decision_weight):
        """
        Determines the final action based on regional modular handlers.
        """
        if current_status:
            if self.leave_handler.evaluate_leave(country_name, decision_weight):
                return "leave"
            return "stay"
        else:
            if self.join_handler.evaluate_join(country_name, decision_weight):
                return "join"
            return "none"

    def get_dynamic_stats(self, members_list, all_countries_data):
        """
        Calculates dynamic regional stats.
        """
        total_pop = 0
        total_gdp = 0
        member_count = len(members_list)
        
        for country_name in members_list:
            data = all_countries_data.get(country_name, {})
            total_pop += data.get('populasi', 0)
            total_gdp += data.get('gdp', 0)
            
        return {
            "member_count": member_count,
            "total_population": total_pop,
            "total_gdp": total_gdp,
            "average_stability": sum([all_countries_data.get(m, {}).get('stabilitas', 50) for m in members_list]) / max(1, member_count)
        }

File: app/frontend/src/app/game/components/2_navigasi_menu/2_navigasi_bawah/5_geopolitik/2_organisasi_internasional/2_organisasi_regional/logic/python/membership_movement/RegionalMovementEngine.ts
import { RegionalJoinHandler } from "./1_join_logic/RegionalJoinHandler";
import { RegionalLeaveHandler } from "./2_leave_logic/RegionalLeaveHandler";

/**
 * RegionalMovementEngine TS Bridge
 */
export class RegionalMovementEngine {
    private joinHandler: RegionalJoinHandler;
    private leaveHandler: RegionalLeaveHandler;

    constructor() {
        this.joinHandler = new RegionalJoinHandler();
        this.leaveHandler = new RegionalLeaveHandler();
    }

    public calculateMovement(countryName: string, currentStatus: boolean, decisionWeight: number): 'join' | 'leave' | 'stay' | 'none' {
        if (currentStatus) {
            if (this.leaveHandler.evaluateLeave(countryName, decisionWeight)) {
                return 'leave';
            }
            return 'stay';
        } else {
            if (this.joinHandler.evaluateJoin(countryName, decisionWeight)) {
                return 'join';
            }
            return 'none';
        }
    }

    public getDynamicStats(membersList: string[], allCountriesData: Record<string, any>) {
        let totalPop = 0;
        let totalGdp = 0;
        const memberCount = membersList.length;

        membersList.forEach(m => {
            const data = allCountriesData[m] || {};
            totalPop += data.populasi || 0;
            totalGdp += data.gdp || 0;
        });

        return {
            memberCount,
            totalPopulation: totalPop,
            totalGdp,
            averageStability: membersList.reduce((acc, m) => acc + (allCountriesData[m]?.stabilitas || 50), 0) / Math.max(1, memberCount)
        };
    }
}
